colearn_keras/keras_learner.py

```diff
@@ -76,13 +76,46 @@ def __init__(self, model: keras.Model,
 
         self.vote_score: float = self.test(self.train_loader)
 
+    def recompile_model(self):
+        """
+        Recompiles the Keras model. This way the optimizer history get erased,
+        which is needed before a new training round, otherwise the outdated history is used.
+        """
+        compile_args = self.model._get_compile_args()  # pylint: disable=protected-access
+        opt_config = self.model.optimizer.get_config()
+
+        if opt_config['name'] == 'Adadelta':
+            compile_args['optimizer'] = keras.optimizers.Adadelta.from_config(opt_config)
+        elif opt_config['name'] == 'Adagrad':
+            compile_args['optimizer'] = keras.optimizers.Adagrad.from_config(opt_config)
+        elif opt_config['name'] == 'Adam':
+            compile_args['optimizer'] = keras.optimizers.Adam.from_config(opt_config)
+        elif opt_config['name'] == 'Adamax':
+            compile_args['optimizer'] = keras.optimizers.Adamax.from_config(opt_config)
+        elif opt_config['name'] == 'Ftrl':
+            compile_args['optimizer'] = keras.optimizers.Ftrl.from_config(opt_config)
+        elif opt_config['name'] == 'Nadam':
+            compile_args['optimizer'] = keras.optimizers.Nadam.from_config(opt_config)
+        elif opt_config['name'] == 'RMSprop':
+            compile_args['optimizer'] = keras.optimizers.RMSprop.from_config(opt_config)
+        elif opt_config['name'] == 'SGD':
+            compile_args['optimizer'] = keras.optimizers.SGD.from_config(opt_config)
+        else:
+            raise NotImplementedError(f'''
+            The history erasing is not implemented yet for {opt_config['name']} optimizer.
+            Please raise an issue ticket.
+            ''')
+        self.model.compile(**compile_args)
+
     def mli_propose_weights(self) -> Weights:
         """
         Trains model on training set and returns new weights after training
         - Current model is reverted to original state after training
         :return: Weights after training
         """
         current_weights = self.mli_get_current_weights()
+        # erase the outdated optimizer memory (momentums mostly)
+        self.recompile_model()
```

Comment:
we always want reset? should it be configurable?

Reply:
I'd say almost always. I add a binary option need_reset_optimizer to both learners, which will have a default value of True. 

This way, all demos can remain untouched, yet still there is a way to disable optimizer resetting.

Reply:
cool